File: lab3.2.go
package main

import (
    "bufio"
    "fmt"
    "math"
    "math/rand"
    "os"
    "strconv"
    "strings"
    "time"
)

var rnd = rand.New(rand.NewSource(time.Now().UnixNano()))

func randInt(a, b int) int {
    return rnd.Intn(b-a+1) + a
}
func randFloat(a, b float64) float64 {
    return a + rnd.Float64()*(b-a)
}

// решето Эратосфена
func prime(n int) []int {
    isPrime := make([]bool, n+1)
    for i := 2; i <= n; i++ {
        isPrime[i] = true
    }
    for i := 2; i*i <= n; i++ {
        if isPrime[i] {
            for j := i * i; j <= n; j += i {
                isPrime[j] = false
            }
        }
    }
    out := make([]int, 0, n/10)
    for i := 2; i <= n; i++ {
        if isPrime[i] {
            out = append(out, i)
        }
    }
    return out
}

// возведение по модулю
func powMod(a, b, mod int) int {
    res := 1
    a %= mod
    for b > 0 {
        if b&1 != 0 {
            res = (res * a) % mod
        }
        a = (a * a) % mod
        b >>= 1
    }
    return res
}

// ---- Miller-Build и тесты ----

type MillerRes struct {
    n int
    q []int
}

// milerBuild: генерирует n=2*m+1 с помощью простых
func millerBuild(primes []int, b int) MillerRes {
    maxInd := 0
    for maxInd < len(primes) && primes[maxInd] < (1<<(b-1)) {
        maxInd++
    }
    m := 1
    qvec := []int{}
    for {
        idx := randInt(0, maxInd-1)
        pr := randInt(1, b-1)
        tmp := 1
        for i := 0; i < pr; i++ {
            tmp *= primes[idx]
        }
        if tmp > 0 && m <= math.MaxInt32/tmp {
            m *= tmp
            qvec = append(qvec, primes[idx])
        }
        if m > (1 << (b - 2)) {
            if m >= (1 << (b - 1)) {
                m = 1
                qvec = []int{}
            } else {
                break
            }
        }
    }
    return MillerRes{2*m + 1, qvec}
}

// тест Миллера — ферма + добавление проверки q
func millerTest(n int, q []int, t int) bool {
    if n < 2 {
        return false
    }
    aVals := make([]int, 0, t)
    for i := 0; i < t && i < n-3; i++ {
        aVals = append(aVals, randInt(2, n-2))
    }
    for _, a := range aVals {
        if powMod(a, n-1, n) != 1 {
            return false
        }
    }
    for _, g := range q {
        ok := false
        for _, a := range aVals {
            if powMod(a, (n-1)/g, n) != 1 {
                ok = true
                break
            }
        }
        if !ok {
            return false
        }
    }
    return true
}

// Miller–Rabin тест
func millerRabin(n, t int) bool {
    if n < 2 || n%2 == 0 {
        return n == 2
    }
    d := n - 1
    s := 0
    for d%2 == 0 {
        d /= 2
        s++
    }
    for i := 0; i < t; i++ {
        a := randInt(2, n-2)
        x := powMod(a, d, n)
        if x != 1 && x != n-1 {
            composite := true
            for r := 1; r < s; r++ {
                x = powMod(x, 2, n)
                if x == n-1 {
                    composite = false
                    break
                }
            }
            if composite {
                return false
            }
        }
    }
    return true
}

// ---- Pocklington ----

type PockRes struct {
    n int
    q []int
}

func pocklingtonBuild(primes []int, b int) PockRes {
    maxInd := 0
    for maxInd < len(primes) && primes[maxInd] < (1<<((b/2)+1)) {
        maxInd++
    }
    m := 1
    qvec := []int{}
    for {
        idx := randInt(0, maxInd-1)
        powcnt := randInt(1, (b/2)+1)
        tmp := int(math.Pow(float64(primes[idx]), float64(powcnt)))
        if tmp > 0 && m <= math.MaxInt32/tmp {
            m *= tmp
            qvec = append(qvec, primes[idx])
        }
        if m > (1 << (b / 2)) {
            if m >= (1 << ((b/2)+1)) {
                m = 1
                qvec = nil
            } else {
                break
            }
        }
    }
    R := 0
    for {
        low := (1 << ((b/2) - 1)) + 1
        high := (1 << (b / 2)) - 1
        R = randInt(low, high)
        if R%2 == 0 {
            break
        }
    }
    return PockRes{R*m + 1, qvec}
}

func pocklingtonTest(n int, F []int, rounds int) bool {
    for i := 0; i < rounds; i++ {
        a := randInt(2, n-2)
        if powMod(a, n-1, n) != 1 {
            return false
        }
        passed := false
        for _, q := range F {
            if powMod(a, (n-1)/q, n) == 1 {
                passed = true
                break
            }
        }
        if !passed {
            return true
        }
    }
    return false
}

// ---- GOST generator ----

func gostBuild(primes []int, b int) int {
    maxInd := 0
    for maxInd < len(primes) && primes[maxInd] < (1 << (b / 2)) {
        maxInd++
    }
    q := primes[randInt(0, maxInd-1)]
    for {
        ep := randFloat(0, 1)
        nVal := (math.Pow(2, float64(b-1))/float64(q) +
            math.Pow(2, float64(b-1))*ep/float64(q))
        nn := int(math.Ceil(nVal))
        if nn%2 != 0 {
            nn++
        }
        for u := 0; u < 1000; u += 2 {
            p := (nn+u)*q + 1
            if p > (1 << b) {
                break
            }
            if powMod(2, p-1, p) == 1 && powMod(2, nn+u, p) != 1 {
                return p
            }
        }
    }
}

// количество итераций нужных для вероятности ошибки
func calculateIterations(errorProb float64) int {
    it := 0
    curr := 1.0
    for curr > errorProb {
        curr /= 4.0
        it++
    }
    return it
}

func main() {
    primes := prime(500)

    fmt.Print("Введите размер чисел в битах (7-31): ")
    reader := bufio.NewReader(os.Stdin)
    line, _ := reader.ReadString('\n')
    line = strings.TrimSpace(line)
    b, err := strconv.Atoi(line)
    if err != nil || b < 2 || b > 31 {
        fmt.Println("Ошибка: некорректный размер!")
        return
    }

    const tErr = 0.1
    iters := calculateIterations(tErr)
    fmt.Printf("Используем %d итераций теста для вероятности ошибки <= %v\n", iters, tErr)

    type resT struct {
        val, k int
    }

    // Метод Миллера
    {
        var results []resT
        k := 0
        for len(results) < 10 {
            mr := millerBuild(primes, b)
            n := mr.n
            found := false
            for _, r := range results {
                if r.val == n {
                    found = true
                    break
                }
            }
            if found {
                k++
                continue
            }
            if millerTest(n, mr.q, iters) {
                results = append(results, resT{n, k})
                k = 0
            } else {
                k++
            }
        }
        fmt.Println("\nМетод Миллера:")
        fmt.Println("----------------------------------------")
        fmt.Println("|  №  |   Число   | Проверка |    k   |")
        for i, r := range results {
            fmt.Printf("| %3d | %9d |    +     | %6d |\n", i+1, r.val, r.k)
        }
    }

    // метод Поклингтона
    {
        var results []resT
        k := 0
        for len(results) < 10 {
            pr := pocklingtonBuild(primes, b)
            n := pr.n
            found := false
            for _, r := range results {
                if r.val == n {
                    found = true
                    break
                }
            }
            if found {
                k++
                continue
            }
            if pocklingtonTest(n, pr.q, iters) {
                results = append(results, resT{n, k})
                k = 0
            } else {
                k++
            }
        }
        fmt.Println("\nМетод Поклингтона:")
        fmt.Println("----------------------------------------")
        fmt.Println("|  №  |   Число   | Проверка |    k   |")
        for i, r := range results {
            fmt.Printf("| %3d | %9d |    +     | %6d |\n", i+1, r.val, r.k)
        }
    }

    // ГОСТ
    {
        var results []resT
        k := 0
        for len(results) < 10 {
            g := gostBuild(primes, b)
            found := false
            for _, r := range results {
                if r.val == g {
                    found = true
                    break
                }
            }
            if found {
                k++
                continue
            }
            if millerRabin(g, iters) {
                results = append(results, resT{g, k})
                k = 0
            } else {
                k++
            }
        }
        fmt.Println("\nМетод ГОСТ:")
        fmt.Println("----------------------------------------")
        fmt.Println("|  №  |   Число   | Проверка |    k   |")
        for i, r := range results {
            fmt.Printf("| %3d | %9d |    +     | %6d |\n", i+1, r.val, r.k)
        }
    }
}

File: lab3.2.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <algorithm>
#include <iomanip>

using namespace std;


mt19937 mRand(random_device{}());
int rand(int a, int b) {
    return uniform_int_distribution<int>(a, b)(mRand);
}
double rd(double a, double b) {
    return uniform_real_distribution<double>(a, b)(mRand);
}

//генерация простых чисел до n с помощью решета Эратосфена
vector<int> prime(int n)
{
    vector<bool> isPrime(n+1, true);

    //путь до корень n
    for(int i =2; i* i <= n; i++)
    if(isPrime[i]){
        for(int j = i*i; j<=n; j+=i)
        {
            isPrime[j] = false;
        }
    }

    vector<int> primes;
    for(int i=2; i<=n;i++)
    {
        if(isPrime[i])
        {
            primes.push_back(i);
        }
    }
    return primes;
}
//возведение в степерь по модулю
long powMod(long a, long b, int mod)
{
    long res = 1;
    a %= mod;
    while(b >0)
    {
        if(b & 1) //проверка бита
            res = (res*a)%mod; //умножаем на текущее значене

        a = (a * a )%mod;
        b >>= 1; //сдвигаем b вправо на 1 бит
    }
    return res;
}

//генерация миллера (числа n вида 2*m+1)
pair<int, vector<int>> millerBuild(vector<int>& prime, int b)
{
    int maxInd = 0; 
    while(maxInd < prime.size() && prime[maxInd] < (1 << (b - 1)))
        maxInd++; //определяем верхнюю границу индексов простых чисел
    
    int m=1; //произведение степеней простых чисел
    vector<int> p; 

    while(true) //цикл генерации
    {
        int idx = rand(0, maxInd -1); //случайный индекс простого числа
        int pr = rand(1,b-1); //случ степень
        long tmp = 1;
        for(int i=0; i < pr; i++)
            tmp *= prime[idx];
        
        if(tmp > 0 && m <= INT32_MAX / tmp) //проверка на отсутствие переполения числа
        {
            m *= tmp; //умножаем на вычисденныъ степенной множитель
            p.push_back(prime[idx]);
        }
        //проверка на превышение нижней и верхней границы
        if(m> (1<< (b - 2))){ // 2^b-2
            if(m >= (1 << (b - 1))){ 
                m=1;
                p.clear(); //сбрасываем, т.к m не подходит
            }else   
                break;
        }
       }
       return {2 * m + 1, p}; //возвращаем готовую пару
    }

//тест Миллера на простоту
bool miller_test(int n, vector<int> q, int t)
{
    if (n<2)
        return false;
    
    vector<int> a_val;
    for(int i=0; i<min(t,n-3);i++)
    {
        a_val.push_back(rand(2,n-2)); //заносим случайные числа 
        //главное чтоб не больше чисел чем в диапазоне
    }
    //проверка a^(n-1) mod n ==1
    for(int a : a_val)
    { //тест ферма a^n-1 = 1(mod n)
        if(powMod(a,n-1,n)!=1){
            return false; //число составное если тест не выполнился
        }
    }

    for(int g : q)
    {
        bool isPrime = false; //наличие подходящего числа
        for(int v : a_val)
        {
            if (powMod(v,(n-1) / g , n)!= 1) //проверка Теоремы ферма
        // v^(n-1)/g если будет показан a mod(n) не делитель для всех простых g, то v равно самому n-1
            {
                isPrime = true;
                break;
            }
        }
        if (!isPrime){
            return false;
        }
    }
    return true;
}

bool miller_rabin(int n, int t)
{
    //проверка чётных и малых чисел
    if(n < 2 || n% 2 ==0)
        return n==2; //вернёт true если двойка


    int d = n - 1; // нечётное число
    int s = 0; //степень двойки
    while (d % 2 == 0) {
            d /= 2;
            s++;
        }
    
    
    for (int i = 0; i < t; ++i) {
        int a = rand(2, n - 2);// выбираем случайное число и вычисляем x = a^d
        long x = powMod(a, d, n);
    
    
        if (x != 1 && x != n - 1){ //полвеояем остальные члены последовательности
            
    
        bool isComposit = true; //составность числа
        //проверка на простое
        for (int r = 1; r < s; ++r) { 
            x = powMod(x, 2, n); //поиск по степени двойки
            if (x == n - 1) {
                isComposit = false;
                break;
            }
        }
        //если x не становится n-1 ни на одном шаге, то число составное
            if (isComposit) 
                return false;
        }
    }
    return true;
}


pair<int, vector<int>> pocklingtonBuild(const vector<int>& primes, int b) {
    int maxInd = 0;
    while (maxInd < primes.size() && primes[maxInd] < (1 << ((b / 2) + 1)))//верхняя граница
        maxInd++;


    int m = 1; //произведение простых множителей
    vector<int> q;


    while (true) {
        int num = rand(0, maxInd - 1);
        int power = rand(1, (b / 2) + 1);
        long temp = pow(primes[num], power); //простое число в степени


        if (temp > 0 && m <= INT32_MAX / temp) {
            m *= temp;
            q.push_back(primes[num]);
        }

//строится число, разрер которого на 1 бит больше половины требуемого размера для простого числа
        if (m > (1 << (b / 2))) {
            if (m >= (1 << ((b / 2) + 1))) {
                m = 1;
                q.clear();
            } else break;
        }
    }
    

    int R;
    //генерирует число, пока не попадётся четный вариант 
    do { //генерируем число размером b/2 то есть (если b = 16 то 8 бит от 128 до 255)
        //сдвигаем верхнюю и нижнюю границу диапазона и генерируем число в этом промежутке
        R = rand((1 << ((b / 2) - 1)) + 1, (1 << (b / 2)) - 1);
    } while (R % 2 != 0);
    //возвращем число собранное по формуле
    return {R * m + 1, q};
}

bool pocklington_test(int n, const vector<int>& F, int rounds) {

    for (int i = 0; i < rounds; ++i) {
        int a = rand(2, n - 2);
        if (powMod(a, n - 1, n) != 1) //проверка по малой теореме ферма
            return false;//если a^n-1 mod n !=1 то оно составное

        bool passed = false;
        for (int q : F) { //перебор простых делителей n-1
            if (powMod(a, (n - 1) / q, n) == 1) { //условие от обратного т.к
            //ищем число a для которого условие a^(n-1)/q != 1 выполняется для всех делителей
                passed = true; 
                break;
            }
        }
        if (!passed) 
            return true; //n — вероятно простое
    }
    return false; //не прошло тест
}
//гост, переход от меньшего простого к большему
int gostBuild(const vector<int>& primes, int b) {

    int maxInd = 0;
    while (maxInd < primes.size() && primes[maxInd] < (1 << (b / 2)))
        maxInd++; //до какого индекса в primes находятся подходящие простые числа
        
    int q = primes[rand(0, maxInd - 1)]; //генерация базового простое число

        // Вместо 1000 попыток, ищем пока не найдем
        while (true) {
        double ep = rd(0,1); //кси
        //вычисление n из формулы 2^(t-1)/q + 2^(t-1)ξ)/q
        double n_val =( pow(2, b-1)/q) + (pow(2, b - 1) * ep/q);
        int n = ceil(n_val);//округляет в большую сторону
        if (n % 2 != 0) 
            n++;

    
        for (int u = 0; u < 1000; u += 2) {//перебираем четные числа пока не получим простое число
            int p = (n + u) * q + 1; //условное значение на простое число
                if (p > (1 << b)){ //проверка на превышение числа заданной битности (p>2^b)
                    break;
                }
                //по теореме Диемитко проверку
                // 2^p-1 == 1 mod p и 2^N+u != 1 mod p
                if (powMod(2, p - 1, p) == 1 && powMod(2, n + u, p) != 1)
                    return p;
            }
        }
    
}
//вычисление количества проводимых тестов на простату, чтобы вероятность ошибки была мала
int calculateIterations(double errorProb) {
    int it =0;
    double currProb = 1.0; //текущая вероятность ошибки
    while(currProb > errorProb)
    {
        currProb /=4.0; //одна итерация = уменьшение вероятности ошибки
        it++;
    }
    return it;
}


int main()
{
    vector<int> prs = prime(500);

    int b;
    cout << "Введите размер чисел в битах (7-31): ";
    cin >> b;

    if (b < 2 || b > 31) {
        cout << "Ошибка: некорректный размер!" << endl;
        return 1;
    }
    const double t_Error = 0.1; //вероятность ошибки
    const int testIt = calculateIterations(t_Error);
    cout << "Используем " << testIt << " итераций теста для вероятности ошибки <= " << t_Error << endl;

    vector<pair<int, int>> res;
    //Метод Миллера
    {
        res.clear();
        int k = 0;

        while (res.size() < 10) {
            auto [n, q] = millerBuild(prs, b); //генерация числа используя список простых
            bool found = false;
            for(auto &p : res){
                if(p.first == n){ //проверка на дубликат
                    found = true;
                    break;
                }
            }
        if (found)
            continue;

        if(miller_test(n,q,testIt)){
            res.push_back({n,k});
            k=0;
        } else
        {
            k++;
        }
    }

        cout << "\nМетод Миллера:\n";
        cout << "----------------------------------------\n";
        cout << "|  №  |   Число   | Проверка |    k   |\n";


        for (size_t i = 0; i < res.size(); ++i) {
            bool check = miller_rabin(res[i].first, testIt); // по миллеру--рабину проверяем массив и насколько серьёзно
            cout << "| " << setw(3) << i + 1 << " | " << setw(9) << res[i].first  << setw(9) 
                 << (check ? '+' : '-') << setw(6) << res[i].second << " |\n";
        }
    }


    //метод поклингтона
    {
        res.clear();

        int k = 0;

        while (res.size() < 10) {
            auto [n, q] = pocklingtonBuild(prs, b);
            bool found = false;

            for(auto &p : res)
            {
                if(p.first == n){
                    found == true;
                    break;
                }
            }
            if (found)
                continue;
            if (pocklington_test(n,q,testIt))
            {
                res.push_back({n,k});
                k=0;
            }else{
                k++;
            }
        }


        cout << "\nМетод Поклингтона:\n";
        cout << "----------------------------------------\n";
        cout << "|  №  |   Число   | Проверка |    k   |\n";


        for (size_t i = 0; i < res.size(); ++i) {
            bool check = miller_rabin(res[i].first, testIt);
            cout << "| " << setw(3) << i + 1 << " | " << setw(9) << res[i].first << setw(9) 
                 << (check ? '+' : '-') << setw(6) << res[i].second << " |\n";
        }
    }


    //гост
    {
        res.clear();
        int k = 0;


        while (res.size() < 10) {
            int g = gostBuild(prs, b);
            bool found = false;
            for(auto &p : res)
            {
                if(p.first== g){
                    found == true;
                    break;
                }
            }

            if(found)
                continue;
            if(miller_rabin(g,testIt))
            { //финальная проверка методом Миллера-рабина
                res.push_back({g,k});
                k=0;
            }else
                k++;
        }

        cout << "\nМетод ГОСТ:\n";
        cout << "----------------------------------------\n";
        cout << "|  №  |   Число   | Проверка |    k   |\n";
        // в госте K - проходит проверку с первой попытки и по этому оно всегда равно 0

        for (size_t i = 0; i < res.size(); ++i) {
            bool check = miller_rabin(res[i].first, testIt);
            cout << "| " << setw(3) << i + 1 << " | " << setw(9) << res[i].first << setw(9) 
                 << (check ? '+' : '-') << setw(6) << res[i].second << " |\n";
        }

    }

    return 0;
}
